pkg/plugins/plugins.go

```diff
@@ -0,0 +1,63 @@
+package plugins
+
+// OpenShiftAzureConfig is a new, Red Hat owned type defined in an external project.  It
+// represents a superset of the configuration required for all steps in the plugin process *after*
+// the manifest is validated.
+//
+// OpenShiftAzureConfig is a versioned type.  It is expected that the configuration will change
+// between releases and may require migration steps in the future.  The storage of this type
+// is opaque to the plugins, they only require that it be passed in.  However, it must be retained
+// to allow it to be lifecycled from version to version.
+type OpenShiftAzureConfig struct{}
+
+// ManagedOpenShiftCluster is the external schema that is passed to the RP.  It is defined in
+// acs-engine.  ManagedOpenShiftCluster will reuse many of the lower level objects shared by
+// ManagedCluster but is separated in order to keep the relationship with AKS clear.
+type ManagedOpenShiftCluster struct{}
+
+// Validator interface is the first plugin in the chain.  It is responsible for validation of
+// both ManagedOpenShiftCluster.  ManagedOpenShiftCluster enters this process in a versioned state.
+// It is validated both according to the version and against the previous state of the world.
+// Versioned validation should occur first.  The manifest, if acceptable, may then be converted
+// into the internal representation and validated against the previous state of the world.
+//
+// Since this plugin validates with both a versioned manifest and an unversioned manifest it will
+// be necessary to cycle the plugin along side the api versions.
+//
+// TODO: is the lifecycle here acceptable?  It is how api.validate.go works.  Does this make it harder
+// on the RP calling the method?  Might have to provide an api version aware factory to return the correct
+// implementation.
```

Comment:
Or extend the interface later - Validate20180101(), Validate20190101()...?

Reply:
yuck.  If we take in bytes we can figure out the version and validate against it without changing the interface.  We do it in origin.  Maybe that's more complicated that what we want to start with but I'd like to be able to hide the complexity if possible